### PR TITLE
generalization and vectorization of GL computation

### PR DESCRIPTION
**Pull request by MoiColl**
In this branch I've generalised `sim_allelereadcounts()` for user specified "ploidy" instead of assuming diploid individuals. Furthermore, I've also generalised the genotype likelihood computation for ploidy and vectorised it's computation (creating a new function `allelereadcounts_to_vGL()`). I left the original `allelereadcounts_to_GL` (which assumes diploid individuals and uses for loops instead of vectorised functions) in case we want to compare both performance. I've also created a couple of functions that can be useful to test the code: `read_angsd_gl()` which reads the GL output from angsd and formats it into numpy array; `msqrd()` mean squared root difference between two arrays that can be used to compute differences between two arrays (e.g., for GL calculation comparison).

### CODE REVIEW

File: simGL/simGL.py [hunk 1]
```diff
@@ -15,32 +16,30 @@ def extract_genotype_matrix(data):
...
+                dp = rng.normal(loc = mean_depth/ploidy, scale = std_depth, size=1)[0]
                 if dp > 0:
                     DPh.append(dp)
             return DPh
-    elif type(mean_depth) == np.ndarray and len(mean_depth.shape) == 1 and mean_depth.shape[0] == n_ind and (mean_depth > 0).sum() == n_ind:
+    elif type(mean_depth) == np.ndarray and len(mean_depth.shape) == 1 and mean_depth.shape[0] == n_hap and (mean_depth > 0).sum() == n_hap:
         return mean_depth
     else:
-        sys.exit("Incorrect mean_depth format")
+        raise Exception('Incorrect mean_depth format')
```
Comment (grahamgower) [inline simGL/simGL.py:31]: This should probably raise a https://docs.python.org/3/library/exceptions.html#TypeError here, rather than Exception.

File: simGL/simGL.py [hunk 2]
```diff
@@ -15,32 +16,30 @@ def extract_genotype_matrix(data):
     else:
         sys.exit("Incorrect data format")
 
-def depth_per_haplotype(rng, mean_depth, std_depth, n_ind):
-    if type(mean_depth) == int or type(mean_depth) == float:
+def depth_per_haplotype(rng, mean_depth, std_depth, n_hap, ploidy):
+    if (type(mean_depth) == int or type(mean_depth) == float) and mean_depth > 0.0:
```
Comment (grahamgower) [inline simGL/simGL.py:20]: It's generally better to use `isinstance()` to check types, rather than comparing the types for equality. This is more flexible and robust in a variety of cases (e.g. it works with subclasses).
`if isinstance(mean_depth, (int, float)) and mean+depth > 0:`

File: simGL/simGL.py [hunk 3]
```diff
@@ -91,6 +91,44 @@ def allelereadcounts_to_GL(Rg, e = 0.05):
...
                                   np.power(((e/3)/2 + (e/3)/2), Rg.sum(axis = 2)-Rg[:, :, i]-Rg[:, :, j])))
 
-    GL = np.array(GL)
-    return GL - GL.min(axis = 0)
+    GL = np.array(GL).transpose(1, 2, 0)
+    return GL - GL.min(axis = 2).reshape(GL.shape[0], GL.shape[1], 1)
+
+def allelereadcounts_to_pileup(allelereadcounts, filename = "tmp/reads.pileup"):
+    with open(filename, "w") as out:
+        first_line = True
```
Comment (grahamgower) [inline simGL/simGL.py:99]: Unused variable? The `flake8` python package is good at finding these (among other things).

**Review by grahamgower — CHANGES_REQUESTED** [review]

File: simGL.egg-info/PKG-INFO [hunk 4]
```diff
@@ -0,0 +1,16 @@
+Metadata-Version: 1.1
```
Comment (grahamgower) [inline simGL.egg-info/PKG-INFO:1]: The contents of the simGL.egg-info package should be committed into the repository. It's probably a good idea to add this folder to the `.gitignore` file.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import setuptools
```
Comment (grahamgower) [inline setup.py:1]: Nice. It's probably better to start with a `setup.cfg` for new packages though, because using setup.py is deprecated.  You can mostly just translate the lines in the setup.py to analogous lines in setup.cfg. E.g. https://github.com/grahamgower/demesdraw/blob/main/setup.cfg
You'll also need a pyproject.toml file. Some of the details of how things all fit together can be confusing, but there are lots of examples in other python packages.

File: requirements/test.txt [hunk 6]
```diff
@@ -1,3 +1,5 @@
 msprime==1.1.0
 pytest==6.2.5
 pytest-cov==3.0.0
+tskit==0.4.0
+numpy==1.22.1
```
Comment (grahamgower) [inline requirements/test.txt:5]: I don't think it's necessary to add numpy here too, because its already in the minimal.txt file.

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+setuptools.setup(
+    name="simGL", 
+    version="0.1",
+    author="Moisès Coll Macià, Graham Gower",
+    author_email="moicoll@birc.au.dk",
+    description="Simulate genotype likelihoods from haplotype genotype matrix.",
+    packages=["simGL"],
+    license = "ISC",
+    url = "https://github.com/RacimoLab/simGL",
+    install_requires=["numpy>=1.22","scipy>=1.7"],
```
Comment (grahamgower) [inline setup.py:12]: It's usually not necessary to specify a minimum version for the `install_requires` packages, unless you're explicitly using some feature that doesn't exist in older versions (e.g. the [new array interface in numpy](https://numpy.org/doc/stable/release/1.22.0-notes.html#implementation-of-the-nep-47-adopting-the-array-api-standard))

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    classifiers = [
+        "Development Status :: 2 - Pre-Alpha",
+        "Operating System :: OS Independent",
+        "License :: OSI Approved :: ISC License",
+        "Intended Audience :: Science/Research",
+        "Programming Language :: Python :: 3.9",
```
Comment (grahamgower) [inline setup.py:18]: Do you want to support other python versions? Usually they're listed here too. And then you'll want to add a `python_requires = >=3.something` line, like e.g. in the demesdraw setup.cfg.

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import setuptools
+
+setuptools.setup(
+    name="simGL", 
+    version="0.1",
+    author="Moisès Coll Macià, Graham Gower",
+    author_email="moicoll@birc.au.dk",
```
Comment (grahamgower) [inline setup.py:7]: Be aware that your email address will be farmed by spammers if you put it here. There's no requirement to provide an author_email, but it's up to you.

**Review by grahamgower — APPROVED** [review]
Mostly looks good to me! If you remove the egg-info folder, feel free to commit this and we can follow up with some setup.py/setup.cfg changes later.

**Comment by MoiColl** [comment]
Thanks! Here are some answers to some of your comments: 

1) I didn't realize there was this `simGL.egg-info` directory until I committed. I believe it stems from the `setuptools` packaging process. I'll add it to `.gitignore`

2) I'll change from `setup.py` to `setup.cfg` and `pyproject.toml`. When I started with that file, I saw that you suggested using those two, but Jo Bovy's guide suggests using `setup.py`. Since I found his explanation was very clear, I decided to stick to it. 

3) There is a feature of numpy that I'm using that is only available for version 1.22.1, so I think that for numpy we must specify a minimal version. For the rest, if we don't want to specify a version, you just give its name, right? This should be consistent with the `requirements.txt` file, right?

4) I specified the python version I'm using, but I guess we can test for different python versions. Can this be set up in the tests?

5) Regarding my email, I was actually concerned with what you say. Thanks for pointing this out.

**Comment by MoiColl** [comment]
Have you seen that tests for `ubuntu-18.04, 3.7` and `macos-10.15, 3.7` fail? I think it's due to the numpy version. How can this be solved? does it mean that this package will not be installable for those OS?

**Comment by grahamgower** [comment]
> Thanks! Here are some answers to some of your comments:
> 
> 1. I didn't realize there was this `simGL.egg-info` directory until I committed. I believe it stems from the `setuptools` packaging process. I'll add it to `.gitignore`

Cool. Are you familiar with the idea of "squashing" commits? Its possible to remove the folder, commit the change, then "sqaush" those commits so that the commits that added and then removed the folder don't appear in the history. Don't worry about this if you don't want to spend the time figuring out the details (but it's super useful to know).

I've found this online git book to be very nice.
https://git-scm.com/book/en/v2

> 2. I'll change from `setup.py` to `setup.cfg` and `pyproject.toml`. When I started with that file, I saw that you suggested using those two, but Jo Bovy's guide suggests using `setup.py`. Since I found his explanation was very clear, I decided to stick to it.

No problem. It doesn't matter so much, I guess. When you go to install a package with a setup.py file, the usual way is/was to run `python setup.py something`, but this is deprecated and many of the subcommands are starting to break. If a package only has a setup.cfg, that can't be used, and you can do things like `pip install .` instead (which should also work if there's just a setup.py file).

> 3. There is a feature of numpy that I'm using that is only available for version 1.22.1, so I think that for numpy we must specify a minimal version.

Oh, ok. You've done the right thing then.

> For the rest, if we don't want to specify a version, you just give its name, right?

Yep.

> This should be consistent with the `requirements.txt` file, right?

Not necessarily. The requirements.txt file(s) aren't used when the package gets installed (e.g. with pip install). Different projects use requirements.txt files in different ways. I usually use them to mean "here is a collection of versions that are known to work for developing the package", and then this set of package versions can be used in the CI too. Users get the versions defined by the setup.py/setup.cfg, and these are usually more relaxed in terms of version constraints.

> 4. I specified the python version I'm using, but I guess we can test for different python versions. Can this be set up in the tests?

Yep, we can change the tests to suit any version(s).

> 5. Regarding my email, I was actually concerned with what you say. Thanks for pointing this out.

:+1: 

> Have you seen that tests for `ubuntu-18.04, 3.7` and `macos-10.15, 3.7` fail? I think it's due to the numpy version. How can this be solved? does it mean that this package will not be installable for those OS?

Numpy dropped support for Python 3.7 with numpy 1.22. So the options are: (1) simGL also drops support for Python 3.7, or (2) find an alternative for the new numpy feature(s) that simGL is using.

What is the feature that requires numpy 1.22?

**Comment by MoiColl** [comment]
In the function `sim_allelereadcounts()` I use the following statement:

`arc  = rng.multinomial(DP, err[gmbp])`

which `DP` is an numpy array with shape (sites, individuals), `err` is a two dimension numpy array (4, 4) that corresponds to the allele probability. Example: using the convention that each base pair corresponds to an index position (A = 0, C = 1, G = 2 and T = 3) the first row, `err[0, :]`, corresponds to when the correct allele is "A" and thus, the allele error probabilities are `[1-e, e/3, e/3, e/3]`. With the same logic `err[1, :] = [e/3, 1-e, e/3, e/3]` and so on. Finally `gmbp` is a numpyarray very similar to the `ts.genotype_matrix()`. The only difference is that 0 = ancestral and 1 = derived alleles, here I encoded the base pairs (A = 0, C = 1, G = 2 and T = 3). 

So, basically, what I achieve is to sample from a multinomial distribution the allele read counts with an base-pair specific probability array. Using arrays in both arguments (`n` and `p`) for the `multinomial()` function is only possible in the new numpy version. 

If you think there is a vectorized way that we can achieve that, maybe we can modify this function and be able to use older numpy functions.

**Comment by grahamgower** [comment]
That seems like very reasonable usage of numpy. Lets just drop Python 3.7.
